cmd/kms-rest/startcmd/start.go

```diff
@@ -94,78 +121,117 @@ func createStartCmd(srv server) *cobra.Command {
-	startCmd.Flags().StringArrayP(tlsCACertsFlagName, "", []string{}, tlsCACertsFlagUsage)
+	for _, handler := range kmsService.GetOperations() {
+		router.HandleFunc(handler.Path(), handler.Handle()).Methods(handler.Method())
+	}
+
+	log.Infof("starting KMS service on host %s", parameters.hostURL)
+
+	return srv.ListenAndServe(parameters.hostURL, constructCORSHandler(router))
 }
 
-func startKmsService(parameters *kmsRestParameters, srv server) error {
-	rootCAs, err := tlsutils.GetCertPool(parameters.tlsSystemCertPool, parameters.tlsCACerts)
+func createKeystoreProvider() (provider.Provider, error) {
+	c, err := tinkcrypto.New()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	log.Debugf("root ca's %v", rootCAs)
+	// keystore storage is used for storing metadata about keystore and associated keys
+	keystoreStorageProvider := memstore.NewProvider()
+	// kms storage is used for storing keys in a secure manner
+	kmsStorageProvider := ariesmemstorage.NewProvider()
 
-	router := mux.NewRouter()
+	return keystoreProvider{
+		storageProvider: keystoreStorageProvider,
+		kmsCreator:      prepareKMSCreator(kmsStorageProvider),
+		crypto:          c,
+	}, nil
+}
+
+func prepareKMSCreator(kmsStorageProvider ariesstorage.Provider) provider.KMSCreator {
+	return func(keystoreID string) (kms.KeyManager, error) {
+		masterKeyReader, err := prepareMasterKeyReader(kmsStorageProvider)
+		if err != nil {
+			return nil, err
+		}
 
-	// health check
-	router.HandleFunc(healthCheckEndpoint, healthCheckHandler).Methods(http.MethodGet)
+		// TODO: Implement support for masterkey lock (https://github.com/trustbloc/hub-kms/issues/17)
+		secretLockService, err := local.NewService(masterKeyReader, nil)
+		if err != nil {
+			return nil, err
+		}
 
-	// key server
-	// TODO: Replace mock with real implementation (https://github.com/trustbloc/hub-kms/issues/12)
-	keyServer := restapi.New(mock.NewProvider())
-	handlers := keyServer.GetOperations()
+		kmsProv := kmsProvider{
+			storageProvider: kmsStorageProvider,
+			secretLock:      secretLockService,
+		}
 
-	for _, handler := range handlers {
-		router.HandleFunc(handler.Path(), handler.Handle()).Methods(handler.Method())
+		keyURI := fmt.Sprintf(masterKeyURI, keystoreID)
+
+		localKMS, err := localkms.New(keyURI, kmsProv)
+		if err != nil {
+			return nil, err
+		}
+
+		return localKMS, nil
 	}
+}
 
-	log.Infof("starting kms rest server on host %s", parameters.hostURL)
+func prepareMasterKeyReader(kmsStorageProv ariesstorage.Provider) (*bytes.Reader, error) {
+	masterKeyStore, err := kmsStorageProv.OpenStore(masterKeyStoreName)
+	if err != nil {
+		return nil, err
+	}
 
-	return srv.ListenAndServe(parameters.hostURL, constructCORSHandler(router))
+	masterKey, err := masterKeyStore.Get(masterKeyDBKeyName)
+	if err != nil {
+		if errors.Is(err, ariesstorage.ErrDataNotFound) {
+			masterKeyContent := random.GetRandomBytes(keySize)
+			masterKey = []byte(base64.URLEncoding.EncodeToString(masterKeyContent))
```

Comment:
fyi, this line sets an `unencrypted` master key to be stored in the line below.

when master lock is created in #17, you need to update this line to set the `encrypted` master key in the store

Reply:
yep, I noticed that already - without updating that line the following error happens: "cipher: message authentication failed"

thank you @Baha-sk for clarifying that part